text/0000-distribute-to-ecr.md

```diff
@@ -0,0 +1,96 @@
+
+Second, working within a single cloud provider ecosystem allows users to take advantage of idiomatic access control mechanisms such as [node-identity][node-identity] or [workload-identity][workload-identity] based registry authentication. While this is probably unnecessary given Paketo images are public and our Dockerhub account should be exempt from rate limiting, hosting images in ECR removes any ambiguity, allowing users to work entirely within a comfortable toolchain.
+
+## Detailed Explanation
+
+The project has already been given an AWS organization funded by the CFF and we have requested the [custom alias][registry-alias] `paketo-buildpacks` for our public registry. We should take the additional step of contacting support and verifying this account so that our repositories will have a verified badge on the ECR Public Gallery (click on any [heroku buildpack][heroku-gallery] repository in the gallery for an example).
+
+A copy of all image artifacts including releases of buildpack, builder, and stack images should be published there in addition to Dockerhub.
+
+Examples:
+- `public.ecr.aws/paketo-buildpacks/builder:base`
+- `public.ecr.aws/paketo-buildpacks/builder:0.3.12-base`
+- `public.ecr.aws/paketo-buildpacks/java:latest`
+- `public.ecr.aws/paketo-buildpacks/run-jammy-tiny:latest`
+
+### Costs
+Given we will be using public repositories, data transfer in is free. Data transfer out without an AWS account is free but limited by source IP. Data transfer out is free within AWS and up to 5TB per months outside of AWS.
+
+Assuming we are unlikely to go over the external data transfer limits, our costs would be limited to storage fees of `$0.10` per GB / month (See [ECR pricing][ecr-pricing] for details).
+
+Some very rough, back of the envelope math...
+
+For simplicity assume:
+1. We do not need to pay twice for blobs that are mounted across repos.
+2. The full builder contains a superset of the buildpack + stack blobs.
+3. There are ~600 version of the builder within the retention policy at any given time (looks like we have published 559 to Dockerhub so far).
+4. The full builder grows to `2GB` (It is currently `1.89GB`).
+
+The CFF will pay `600 * 2GB * $0.10/GB = $120` per month for storage. This is well within reason, even if many of these simplifying assumptions turn out to be wrong and the bill is 5x my rough estimate.
+
+## Rationale and Alternatives
+
+### Alternative 1 - Do Nothing
+We could not do this and require AWS users to consume Paketo images from Dockerhub. If a group like the App Runner team wishes to provide Paketo images on ECR to enhance the App Runner experience they could copy images to an ECR public repo.
+
+If we go this route we run the risk of letting a third party fill the vacuum, becoming the default distributors of Paketo buildpacks on ECR. This may present a reputational risk to to the project and a security risk to end users.
+
+### Alternative 2 - Publish to All the Cloud Providers
```

Comment:
+1 - I hadn't considered the angle of favoritism, but I think that's a very valid point.

My first instinct was to think we should just extend this and publish it everywhere, but I don't agree that is a good use of project funds, also where do you draw the line? Do you publish to Joe's discount cloud-hosting service? If not, why? 

To avoid favoritism and to ensure it's worth our resources, I feel like we need to have some objective criteria established to define if we will or won't publish to a particular vendor. In this case, if Amazon had offered to comp the costs for ECR, that would be different. We could use that as a requirement for other vendors that want us to publish images to their registries. I'm not sure what other requirements we could use? We don't have usage metrics, so we can't use pull-metrics or usage metrics on a given cloud provider. Any thoughts on this point?

Reply:
My understanding was we're doing this to facilitate a better integration with AWS, and that the pertinent AWS team had specifically requested that we do this. This doesn't feel like favoritism to me as much as "meeting an integrator's needs when they tell us about them." @ekcasey can you shed light on how this feature request came about?

 Maybe it's enough to add a note to this RFC that we'd consider doing something similar for other cloud providers to facilitate integrations if they express desire and have users that would benefit.

Reply:
@ForestEckhardt @dmikusa-pivotal @fg-j

Added a [section](https://github.com/paketo-buildpacks/rfcs/blob/distribute-to-ecr/text/0000-distribute-to-ecr.md#other-cloud-providers). Lmk if you think this adequately addresses the above concerns.

We definitely don't want to be in the business of favoring some cloud over others. But I want to balance that against the costs in IaaS money and labor/reputational costs of walking it back if for whatever reason we don't find that the ECR rollout was worthwhile.

Reply:
I'm OK with what you've added. I would like to have something more concrete, but that's a hard problem. 

The last thing I would add is that I feel like there is an opportunity here to start getting some metrics about usage. Should we/can we require something like this from potential integrations opportunities? I don't think we need a ton of information, but if we can get high level, anonymous metrics like the number of downloads, the number of unique users (if a user is logged in when they download), etc... This would help us understand buildapck usage patterns better, which is a big win for us, it would also help to justify the work and expense, objectively.

Reply:
@ekcasey - Bump. Last question, I promise. ^^^

Reply:
@dmikusa-pivotal from a quick perusal of the documentation it looks like [ECR offers RepositoryPullCount and not much else](https://docs.aws.amazon.com/AmazonECR/latest/userguide/ecr-repository-metrics.html#available_cloudwatch_metrics). This is similar to what we can currently get from Dockerhub (pull count and star count). Historically it's been hard to know how to interpret our Dockerhub metrics without data about unique users (e.g. a single CI/CD pipeline that runs frequently can generate a lot of pulls).

While I would like to have better metrics I don't know that I feel comfortable mkaing it a requirement for new registries given:
1. We have limited ability to influence what is available
1. AFAIK our current default registry doesn't provide info like the number of unique users

Reply:
Acknowledged.